common/log.h

```diff
@@ -0,0 +1,658 @@
+//  does the same as LOG
+//  and
+//  simultaneously writes stderr.
+//
+// Secondary target can be changed just like LOG_TARGET
+//  by defining LOG_TEE_TARGET
+//
+#ifndef _WIN32
+/**/#define LOG_TEE(...) _LOG_TEE(__VA_ARGS__, "")
+#else
+/**/#define LOG_TEE(str, ...) _LOG_TEE("%s" str, "", __VA_ARGS__, "")
+#endif
+
+// LOG macro variants with auto endline.
+#ifndef _WIN32
+/**/#define LOGLN(...) _LOG(__VA_ARGS__, "\n")
+/**/#define LOG_TEELN(...) _LOG_TEE(__VA_ARGS__, "\n")
+#else
+/**/#define LOGLN(str, ...) _LOG("%s" str, "", __VA_ARGS__, "\n")
+/**/#define LOG_TEELN(str, ...) _LOG_TEE("%s" str, "", __VA_ARGS__, "\n")
+#endif
+
+// INTERNAL, DO NOT USE
+inline FILE *_log_handler1(bool change = false, LogTriState disable = LogTriStateSame, std::string filename = LOG_DEFAULT_FILE_NAME, FILE *target = nullptr)
+{
+    static bool _initialized{false};
+    static bool _disabled{(filename.empty() && target == nullptr)};
+    static std::string log_current_filename{filename};
+    static FILE *log_current_target{target};
+    static FILE *logfile = nullptr;
+
+    if (change)
+#ifndef _WIN32
+        [[unlikely]]
+#endif
+    {
+        if (disable == LogTriStateTrue)
+        {
+            // Disable primary target
+            _disabled = true;
+        }
+        // If previously disabled, only enable, and keep previous target
+        else if (disable == LogTriStateFalse)
+        {
+            _disabled = false;
+        }
+        // Otherwise, process the arguments
+        else
+        {
+            if (log_current_filename.compare(filename) != 0)
+            {
+                _initialized = false;
+            }
+
+            if (log_current_target != target)
+            {
+                _initialized = false;
+            }
+        }
+    }
+
+    if (_initialized)
+#ifndef _WIN32
+        [[likely]]
+#endif
+    {
+        if (_disabled)
+        {
+            // Log is disabled
+            return nullptr;
+        }
+        else
+        {
+            // with fallback in case something went wrong
+            return logfile ? logfile : stderr;
+        }
+    }
+    else
+    {
+        if (target != nullptr)
+        {
+            if (logfile != nullptr && logfile != stdout && logfile != stderr)
+            {
+                fclose(logfile);
+            }
+
+            log_current_filename = LOG_DEFAULT_FILE_NAME;
+            log_current_target = target;
+
+            logfile = target;
+        }
+        else
+        {
+            if (log_current_filename.compare(filename) != 0)
+#ifndef _WIN32
+                [[likely]]
+#endif
+            {
+                if (logfile != nullptr && logfile != stdout && logfile != stderr)
+                {
+                    fclose(logfile);
+                }
+
+                logfile = nullptr;
+                logfile = fopen(filename.c_str(), "a");
+            }
+            else
+            {
+                logfile = fopen(filename.c_str(), "a");
```

Comment:
This happens in both the true and false cases, why not place it after the 'if' statement? (And why are we setting logfile to nullptr above before setting it to something else?)

Reply:
> And why are we setting logfile to nullptr above before setting it to something else?

My habits from working with C and embedded MCU stuff :) I removed it as it's not necessary in final code.